#### Name : SINU S MARIAM
#### Designation : Machine Learning Intern  
#### Organization : COGNIFYZ TECHNOLOGIES
#### Ref : CTI/A1/C41852
#### Batch Date: 23/06/2024 to 23/09/2024 
#### Task Date : 23/07/2024 to 23/08/2024
#### Second Month Task  - RAILWAY INFORMATION PROCESSING

### Level 1: Data Exploration and Preprocessing
#### Task 1.1: Load and Inspect Data
* Load the dataset and display the first 10 rows.
* Understand the basic structure of the data, including data types and missing values.

In [1]:
#importing necessary libraries 
import numpy as np
import pandas as pd
import warnings 
warnings.filterwarnings('ignore')
#importing libraries for visualisation
import matplotlib.pyplot as plt
from matplotlib import style
import seaborn as sns

In [2]:
#importing Data
data_frame = pd.read_csv('C:/Users/sinus/OneDrive/Documents/COGNIFYZ/secondmonth/Railway_info.csv')

####  Performing descriptive analysis. Understand the variables and their corresponding values. 

In [3]:
# Understanding the dimensions of train data
data_frame.shape

(11113, 5)

In [4]:
# Understanding the Data Variables
data_frame.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11113 entries, 0 to 11112
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Train_No                  11113 non-null  int64 
 1   Train_Name                11113 non-null  object
 2   Source_Station_Name       11113 non-null  object
 3   Destination_Station_Name  11113 non-null  object
 4   days                      11113 non-null  object
dtypes: int64(1), object(4)
memory usage: 434.2+ KB


In [5]:
#Displaying columns 
data_frame.columns

Index(['Train_No', 'Train_Name', 'Source_Station_Name',
       'Destination_Station_Name', 'days'],
      dtype='object')

In [6]:
# Show the top 10 Rows of data
data_frame.head(10)

Train_No    Train_Name           Source_Station_Name  \
0       107  SWV-MAO-VLNK               SAWANTWADI ROAD   
1       108  VLNK-MAO-SWV                   MADGOAN JN.   
2       128  MAO-KOP SPEC                   MADGOAN JN.   
3       290  PALACE ON WH             DELHI-SAFDAR JANG   
4       401  BSB BHARATDA                    AURANGABAD   
5       421  LKO-SVDK FTR                   LUCKNOW JN.   
6       422  SVDK-LKO FTR  SHRI MATA VAISHNO DEVI KATRA   
7       477  FTR TRAIN NO                         SIRSA   
8       502  RJPB-UMB FTR        RAJENDRANAGAR TERMINAL   
9       504  PNBE-BTI FTR                     PATNA JN.   

             Destination_Station_Name       days  
0                         MADGOAN JN.   Saturday  
1                     SAWANTWADI ROAD     Friday  
2  CHHATRAPATI SHAHU MAHARAJ TERMINUS     Friday  
3                   DELHI-SAFDAR JANG  Wednesday  
4                        VARANASI JN.   Saturday  
5        SHRI MATA VAISHNO DEVI KATRA    Tuesday  
6                         LUCKNOW JN.     Monday  
7                               SIRSA     Sunday  
8                     AMBALA CANTT JN     Monday  
9                         BATHINDA JN  Wednesday

In [7]:
# Checking for any Duplicated Entries
data_frame.duplicated().sum()

0

In [8]:
#Detecting missing values
data_frame.isna().sum()

Train_No                    0
Train_Name                  0
Source_Station_Name         0
Destination_Station_Name    0
days                        0
dtype: int64

#### Task 1.2: Basic Statistics
##### Calculate basic statistics such as:
* The number of trains.
* The count of unique source stations and destination stations.
* Find the most common source and destination stations.

In [9]:
data_frame['Train_No'].nunique()

11113

In [10]:
data_frame['Train_Name'].nunique()

7580

* There are 11113 Trains No's and 7580 different Train names in dataset

In [11]:
data_frame['Source_Station_Name'].nunique()

921

In [12]:
data_frame['Destination_Station_Name'].nunique()

924

* There are 921 unique source stations and 924 unique destination stations.

In [13]:
data_frame['Source_Station_Name'].value_counts()

CST-MUMBAI         513
SEALDAH            372
CHENNAI BEACH      339
HOWRAH JN.         338
KALYAN JN          285
                  ... 
MANENDRAGARH         1
LANJI GARH ROAD      1
FURKATING JN.        1
TALCHER              1
KOLAGHAT             1
Name: Source_Station_Name, Length: 921, dtype: int64

In [14]:
data_frame['Destination_Station_Name'].value_counts()

CST-MUMBAI         514
SEALDAH            373
CHENNAI BEACH      342
HOWRAH JN.         337
KALYAN JN          284
                  ... 
KENDUJHAR            1
TORI                 1
GARWA ROAD JN.       1
WADSA                1
LANJI GARH ROAD      1
Name: Destination_Station_Name, Length: 924, dtype: int64

* 'CST-MUMBAI' is the most common station name among source and destination stations

#### Task 1.3: Data Cleaning
* Identify and handle missing values in the dataset using appropriate imputation techniques.
* Standardize the format of station names (e.g., make all names uppercase).

In [15]:
data_frame['days'].value_counts()

Friday        1471
Tuesday       1454
Wednesday     1448
Saturday      1441
Sunday        1432
Thursday      1372
Monday        1342
Fridayd        178
Tuesdayd       174
Sundayd        170
Wednesdayd     164
Mondayd        161
Thursdayd      154
Saturdayd      152
Name: days, dtype: int64

##### * From the values of 'days' , we can understand that the variables needs data cleaning as it contains spelling mistakes

In [16]:
# Remove spelling mistakes from variable 'days'
data_frame['days'].replace({'Mondayd': 'Monday', 'Tuesdayd': 'Tuesday','Wednesdayd': 'Wednesday','Thursdayd': 'Thursday','Fridayd': 'Friday','Saturdayd': 'Saturday','Sundayd': 'Sunday'}, inplace=True)
data_frame['days'].value_counts()


Friday       1649
Tuesday      1628
Wednesday    1612
Sunday       1602
Saturday     1593
Thursday     1526
Monday       1503
Name: days, dtype: int64

In [17]:
# Define function for cleaning data
import re
def Pre_process_text(act_string):
    # remove all values other than alphabets
    new_string= re.sub(r'[^a-zA-Z0-9]'," ", act_string)
    # Lowercasing
    new_string = act_string.upper()
    return new_string

In [18]:
# Text Cleaning for column 'Source_Station_Name'
data_frame['Source_Station_Name']=data_frame['Source_Station_Name'].apply(Pre_process_text)

In [19]:
# Text Cleaning for column Destination_Station_Name' and 'days'
data_frame['Destination_Station_Name']=data_frame['Destination_Station_Name'].apply(Pre_process_text)
data_frame['days']=data_frame['days'].apply(Pre_process_text)

#### Level 2: Feature Engineering and Selection
* Task 2.1: Data Filtering
* Filter the data to show trains that operate on specific days (e.g., only on Saturdays).
* Extract and create a new data frame for trains that start from a specific station.

In [20]:
# Filter the DataFrame for trains that operate on Saturdays
trains_on_saturday= data_frame[data_frame['days'] == 'SATURDAY']

# Display the filtered DataFrame
trains_on_saturday

Train_No    Train_Name Source_Station_Name Destination_Station_Name  \
0           107  SWV-MAO-VLNK     SAWANTWADI ROAD              MADGOAN JN.   
4           401  BSB BHARATDA          AURANGABAD             VARANASI JN.   
21         1196  NGP-KRMI SPL      NAGPUR JN.(CR)                  KARMALI   
28         1706   JBP-BDTS SF            JABALPUR          BANDRA TERMINUS   
45         2834  SRC-RJT SF A     SANTRAGACHI JN.                   RAJKOT   
...         ...           ...                 ...                      ...   
11084     99816  PUNE-LNL EMU            PUNE JN.                  LONAVLA   
11086     99818      PUNE-LNL            PUNE JN.                  LONAVLA   
11094     99826  PUNE-LNL EMU            PUNE JN.                  LONAVLA   
11100     99832  PUNE-LNL EMU        SHIVAJINAGAR                  LONAVLA   
11104     99836  PUNE-TGN EMU            PUNE JN.                  LONAVLA   

           days  
0      SATURDAY  
4      SATURDAY  
21     SATURDAY  
28     SATURDAY  
45     SATURDAY  
...         ...  
11084  SATURDAY  
11086  SATURDAY  
11094  SATURDAY  
11100  SATURDAY  
11104  SATURDAY  

[1593 rows x 5 columns]

In [21]:
# Specify the source station
station_name = 'SHIVAJINAGAR'
# Filter the DataFrame for trains that start from the specific station
trains_station_name = data_frame[data_frame['Source_Station_Name'] == station_name]
# Display the filtered DataFrame
trains_station_name

Train_No    Train_Name Source_Station_Name Destination_Station_Name  \
11082     99814  PUNE-LNL EMU        SHIVAJINAGAR                  LONAVLA   
11100     99832  PUNE-LNL EMU        SHIVAJINAGAR                  LONAVLA   

           days  
11082  THURSDAY  
11100  SATURDAY

### Task 2.2: Feature Generation
* Group the data by source station and count the number of trains originating from each station.
* Find the average number of trains per day for each source station.

In [22]:
# Group by the source station and count the number of trains originating from each station
count_train =data_frame.groupby('Source_Station_Name').size().reset_index(name='Number of Trains')
# Display the resulting DataFrame
count_train

Source_Station_Name  Number of Trains
0          ABHANPUR JN.                 2
1                ABOHAR                 1
2              ABU ROAD                 1
3          ACHHNERA JN.                 1
4              ADILABAD                 5
..                  ...               ...
916           WARASEONI                 1
917          WARDHA JN.                 4
918         WHITE FIELD                 2
919       YAMUNA BRIDGE                 2
920      YESVANTPUR JN.                60

[921 rows x 2 columns]

In [23]:
# Group by source station and day of operation and count the number of trains per day
daily_count = data_frame.groupby(['Source_Station_Name', 'days']).size().reset_index(name='Daily Train Count')
daily_count


Source_Station_Name       days  Daily Train Count
0           ABHANPUR JN.   SATURDAY                  1
1           ABHANPUR JN.    TUESDAY                  1
2                 ABOHAR  WEDNESDAY                  1
3               ABU ROAD    TUESDAY                  1
4           ACHHNERA JN.   SATURDAY                  1
...                  ...        ...                ...
3380      YESVANTPUR JN.   SATURDAY                  8
3381      YESVANTPUR JN.     SUNDAY                  6
3382      YESVANTPUR JN.   THURSDAY                 10
3383      YESVANTPUR JN.    TUESDAY                  9
3384      YESVANTPUR JN.  WEDNESDAY                 11

[3385 rows x 3 columns]

In [24]:
# Group by source station and calculate the average number of trains per day
avg_train_per_day = daily_count.groupby('Source_Station_Name')['Daily Train Count'].mean().reset_index(name='Average Trains per Day')
# Display the resulting DataFrame
avg_train_per_day 

Source_Station_Name  Average Trains per Day
0          ABHANPUR JN.                1.000000
1                ABOHAR                1.000000
2              ABU ROAD                1.000000
3          ACHHNERA JN.                1.000000
4              ADILABAD                1.250000
..                  ...                     ...
916           WARASEONI                1.000000
917          WARDHA JN.                1.000000
918         WHITE FIELD                1.000000
919       YAMUNA BRIDGE                1.000000
920      YESVANTPUR JN.                8.571429

[921 rows x 2 columns]

#### Task 2.3: Feature Enrichment
* Add new columns that categorize trains based on their operating days (e.g.,(e.g., 'Weekday', 'Weekend') and other relevant features to enhance the dataset.

In [25]:
# Define the function to create a weekday or weekend
def weekday_or_not(day):
    if day in ['SATURDAY', 'SUNDAY']:
        return 'Weekend'
    else:
        return 'Weekday'
# Apply the function to create a weekday or weekend
data_frame['WEEKEND_OR_WEEKDAY'] = data_frame['days'].apply(weekday_or_not)
data_frame.to_csv('modified_dataframe.csv', index=False)
data_frame.head(5)

Train_No    Train_Name Source_Station_Name  \
0       107  SWV-MAO-VLNK     SAWANTWADI ROAD   
1       108  VLNK-MAO-SWV         MADGOAN JN.   
2       128  MAO-KOP SPEC         MADGOAN JN.   
3       290  PALACE ON WH   DELHI-SAFDAR JANG   
4       401  BSB BHARATDA          AURANGABAD   

             Destination_Station_Name       days WEEKEND_OR_WEEKDAY  
0                         MADGOAN JN.   SATURDAY            Weekend  
1                     SAWANTWADI ROAD     FRIDAY            Weekday  
2  CHHATRAPATI SHAHU MAHARAJ TERMINUS     FRIDAY            Weekday  
3                   DELHI-SAFDAR JANG  WEDNESDAY            Weekday  
4                        VARANASI JN.   SATURDAY            Weekend

#### Level 3: Model Building and Evaluation
##### Task 3.1: Model Training
* Split the data into training and testing sets.
* Train various machine learning models (e.g., linear regression, decision trees, randomforests) to predict frequency of operation.

In [26]:
# Frequency of Operation
data_frame['Frequency_of_operation'] = data_frame.groupby(['Train_No','Train_Name', 'Source_Station_Name', 'Destination_Station_Name'])['Train_No'].transform('count')
data_frame


Train_No    Train_Name Source_Station_Name  \
0           107  SWV-MAO-VLNK     SAWANTWADI ROAD   
1           108  VLNK-MAO-SWV         MADGOAN JN.   
2           128  MAO-KOP SPEC         MADGOAN JN.   
3           290  PALACE ON WH   DELHI-SAFDAR JANG   
4           401  BSB BHARATDA          AURANGABAD   
...         ...           ...                 ...   
11108     99904  PUNE-TGN EMU            PUNE JN.   
11109     99905           EMU            TALEGAON   
11110     99906           EMU            PUNE JN.   
11111     99907           EMU            TALEGAON   
11112     99908           EMU            PUNE JN.   

                 Destination_Station_Name       days WEEKEND_OR_WEEKDAY  \
0                             MADGOAN JN.   SATURDAY            Weekend   
1                         SAWANTWADI ROAD     FRIDAY            Weekday   
2      CHHATRAPATI SHAHU MAHARAJ TERMINUS     FRIDAY            Weekday   
3                       DELHI-SAFDAR JANG  WEDNESDAY            Weekday   
4                            VARANASI JN.   SATURDAY            Weekend   
...                                   ...        ...                ...   
11108                            TALEGAON    TUESDAY            Weekday   
11109                        SHIVAJINAGAR     MONDAY            Weekday   
11110                            TALEGAON  WEDNESDAY            Weekday   
11111                            PUNE JN.   THURSDAY            Weekday   
11112                            TALEGAON     SUNDAY            Weekend   

       Frequency_of_operation  
0                           1  
1                           1  
2                           1  
3                           1  
4                           1  
...                       ...  
11108                       1  
11109                       1  
11110                       1  
11111                       1  
11112                       1  

[11113 rows x 7 columns]

In [27]:
data_frame['Frequency_of_operation'].value_counts()

1    11113
Name: Frequency_of_operation, dtype: int64

##### LABEL ENCODING

In [28]:
from sklearn.preprocessing import LabelEncoder
feature_columns=['Train_Name','Source_Station_Name','Destination_Station_Name','days','WEEKEND_OR_WEEKDAY']
# Convert categorical columns to numeric using LabelEncoder
for column in feature_columns:
    if data_frame[column].dtype == 'object':
        le = LabelEncoder()
        data_frame[column] = le.fit_transform(data_frame[column])
data_frame

Train_No  Train_Name  Source_Station_Name  Destination_Station_Name  \
0           107        6442                  772                       522   
1           108        7378                  522                       775   
2           128        4040                  522                       206   
3           290        4872                  249                       248   
4           401        1286                   47                       894   
...         ...         ...                  ...                       ...   
11108     99904        5363                  694                       835   
11109     99905        2282                  830                       792   
11110     99906        2282                  694                       835   
11111     99907        2282                  830                       696   
11112     99908        2282                  694                       835   

       days  WEEKEND_OR_WEEKDAY  Frequency_of_operation  
0         2                   1                       1  
1         0                   0                       1  
2         0                   0                       1  
3         6                   0                       1  
4         2                   1                       1  
...     ...                 ...                     ...  
11108     5                   0                       1  
11109     1                   0                       1  
11110     6                   0                       1  
11111     4                   0                       1  
11112     3                   1                       1  

[11113 rows x 7 columns]

##### BUIDING MODEL FOR PREDICTING 'FREQUENCY OF OPERATION'

In [29]:
#First step in building the model is to identify the Feature(Input) variables and Target (Output) variable
features = data_frame.drop(['Frequency_of_operation'], axis=1)
target = data_frame['Frequency_of_operation']

In [30]:
from sklearn.model_selection import train_test_split

In [31]:
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()

In [32]:
# Splitting data for training the model and testing the model
# train size taken as 0.8
X_train, X_test, y_train, y_test = train_test_split(features, target, train_size = .8)
# Dimensions of Train and Test Data sets
print('Train set of features: ', X_train.shape)
print('Test set of features: ', X_test.shape)
print('Target for train: ', y_train.shape)
print('Target for test: ', y_test.shape)

Train set of features:  (8890, 6)
Test set of features:  (2223, 6)
Target for train:  (8890,)
Target for test:  (2223,)


In [33]:
# Fit and transform the training data, transform the test data
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)
X_train_scaled

array([[0.9734171 , 0.27856954, 0.24347826, 0.30552546, 0.83333333,
        0.        ],
       [0.95192433, 0.75758775, 0.47826087, 0.46153846, 0.83333333,
        0.        ],
       [0.15471789, 0.52401689, 0.55652174, 0.43770314, 0.83333333,
        0.        ],
       ...,
       [0.40678951, 0.58168382, 0.21847826, 0.93282774, 0.83333333,
        0.        ],
       [0.12612098, 0.89878596, 0.94673913, 0.21885157, 0.33333333,
        1.        ],
       [0.56872176, 0.65096331, 0.74565217, 0.98266522, 0.16666667,
        0.        ]])

### Building the Models

###  1. Linear Regression Model

In [34]:
# Learn the Linear Regression Model on train data
from sklearn.linear_model import LinearRegression
my_model_LR=LinearRegression()

In [35]:
# Fitting the model in train data set ie the Linear Regression Model should learn from the on Train Data
my_model_LR.fit(X_train_scaled, y_train)

LinearRegression()

In [36]:
# Predicting 'Frequency_of_operation' from Feature Test values
y_pred_LR = my_model_LR.predict(X_test_scaled)
y_pred_LR 

array([1., 1., 1., ..., 1., 1., 1.])

#### Test the Linear Regression model

In [37]:
from sklearn.metrics import accuracy_score,precision_score,recall_score,f1_score

In [38]:
from sklearn.metrics import mean_squared_error

In [39]:
# Compare the predicted values with the true values
mean_squared_error(y_pred_LR, y_test)

0.0

In [40]:
##### Coefficient of Determination or R Squared Value (r2)

In [41]:
from sklearn.metrics import r2_score

In [42]:
# find Coefficient of Determination or R Squared Value (r2)
r2_score(y_test,y_pred_LR)

1.0

In [43]:
accuracy_score(y_test, y_pred_LR)

1.0

In [44]:
recall_score(y_test, y_pred_LR)

1.0

In [45]:
precision_score(y_test, y_pred_LR)

1.0

### 2. Decision Tree model

In [46]:
from sklearn.tree import DecisionTreeRegressor
my_model_DT=DecisionTreeRegressor(max_depth=10,random_state = 0)

In [47]:
# Fitting the DT model in train data set 
my_model_DT=my_model_DT.fit(X_train_scaled, y_train)

In [48]:
# Predicting from Feature Test values
y_pred_DT = my_model_DT.predict(X_test_scaled)
y_pred_DT

array([1., 1., 1., ..., 1., 1., 1.])

#### Testing the Decision Tree Regression Model

In [49]:
# Compare the predicted values with the true values
mean_squared_error(y_pred_DT, y_test)

0.0

In [50]:
# find Coefficient of Determination or R Squared Value (r2)
r2_score(y_test,y_pred_DT)

1.0

In [51]:
accuracy_score(y_test, y_pred_DT)

1.0

In [52]:
precision_score(y_test, y_pred_DT)

1.0

In [53]:
recall_score(y_test, y_pred_DT)

1.0

### 3. Random Forest Classifier Model

In [54]:
from sklearn.ensemble import RandomForestRegressor
my_model_RF=RandomForestRegressor(random_state=10)

In [55]:
# Fitting the Random Forest Classifier model in train data set
my_model_RF=my_model_RF.fit(X_train_scaled, y_train)

In [56]:
# Predicting from Feature Test values
y_pred_RF=my_model_RF.predict(X_test_scaled)
y_pred_RF

array([1., 1., 1., ..., 1., 1., 1.])

In [57]:
#### Testing the Random Forest Classifier Model

In [58]:
# Compare the predicted values with the true values
mean_squared_error(y_pred_RF, y_test)

0.0

In [59]:
# find Coefficient of Determination or R Squared Value (r2)
r2_score(y_test,y_pred_RF)

1.0

In [60]:
accuracy_score(y_test, y_pred_RF)

1.0

In [61]:
precision_score(y_test, y_pred_RF)

1.0

In [62]:
recall_score(y_test, y_pred_RF)

1.0

 * In the given dataset , the frequency of Operation has only one unique value ie '1'.
 * So we can see that 'accuracy', 'precision', 'recall' are '1.0' and the 'mean squared error' value is '0.0' for all the 3 models.

##### * For the further proccessing, Random Forest Regressor model is taken out of 3 models assuming that it has better performance.

#### Level 4: Advanced Model Optimization and Deployment
##### Task 4.1: Hyperparameter Tuning
 
* Perform hyperparameter tuning on the best-performing model using techniques such as Grid Search or Random Search to improve its performance.

In [ ]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import train_test_split, GridSearchCV
rf = RandomForestRegressor(random_state=42)

# hyperparameters grid to search over
param_grid = {
    'n_estimators': [50, 100, 200],          # Number of trees in the forest
    'max_depth': [None, 10, 20, 30],         # Maximum depth of the tree
    'min_samples_split': [2, 5, 10],         # Minimum number of samples required to split an internal node
    'min_samples_leaf': [1, 2, 4]            # Minimum number of samples required to be at a leaf node
}

# GridSearchCV
grid_search = GridSearchCV(estimator=rf, param_grid=param_grid, cv=5, n_jobs=-1, verbose=2, scoring='neg_mean_squared_error')
# grid search on the training data
grid_search.fit(X_train_scaled, y_train)
#  Extract of  best model 
best_rf = grid_search.best_estimator_
y_pred_RFG = best_rf.predict(X_test_scaled)
# Evaluate
mse = mean_squared_error(y_pred_RFG, y_test)
rmse = mse ** 0.5
print(f"Best Hyperparameters: {grid_search.best_params_}")
print(f"Mean Squared Error: {mse}")
print(f"Root Mean Squared Error: {rmse}")
# Save the best model for future use
import joblib

Fitting 5 folds for each of 108 candidates, totalling 540 fits


exception calling callback for <Future at 0x266f3cd2c50 state=finished raised TerminatedWorkerError>
Traceback (most recent call last):
  File "C:\Users\sinus\anaconda3\Lib\site-packages\joblib\externals\loky\_base.py", line 26, in _invoke_callbacks
    callback(self)
  File "C:\Users\sinus\anaconda3\Lib\site-packages\joblib\parallel.py", line 385, in __call__
    self.parallel.dispatch_next()
  File "C:\Users\sinus\anaconda3\Lib\site-packages\joblib\parallel.py", line 834, in dispatch_next
    if not self.dispatch_one_batch(self._original_iterator):
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\sinus\anaconda3\Lib\site-packages\joblib\parallel.py", line 901, in dispatch_one_batch
    self._dispatch(tasks)
  File "C:\Users\sinus\anaconda3\Lib\site-packages\joblib\parallel.py", line 819, in _dispatch
    job = self._backend.apply_async(batch, callback=cb)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\sinus\anaconda3\Lib\site-pack

#### Feature Importance Plot

In [ ]:
# Re-importing necessary libraries and re-preprocessing the data
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.ensemble import RandomForestClassifier


data_frame = pd.read_csv('C:/Users/sinus/OneDrive/Documents/COGNIFYZ/secondmonth/Railway_info.csv')
data_frame.info()
data_frame['days'].replace({'Mondayd': 'Monday', 'Tuesdayd': 'Tuesday','Wednesdayd': 'Wednesday','Thursdayd': 'Thursday','Fridayd': 'Friday','Saturdayd': 'Saturday','Sundayd': 'Sunday'}, inplace=True)

data_frame = data_frame.drop('Train_No', axis=1)
# Convert categorical columns to numeric using LabelEncoder
for column in data_frame.columns:
    if data_frame[column].dtype == 'object':
        le = LabelEncoder()
        data_frame[column] = le.fit_transform(data_frame[column])

data_frame['Frequency_of_operation'] = data_frame.groupby(['Train_Name', 'Source_Station_Name', 'Destination_Station_Name','days'])['Train_Name'].transform('count')

# Split the data into features and target
X_fi = data_frame.drop('Frequency_of_operation', axis=1)
y_fi = data_frame['Frequency_of_operation']

# Train the Random Forest Classifier
rf_fi = RandomForestClassifier(random_state=50)
rf_fi.fit(X_fi, y_fi)

# Get feature importances and sort them
importances = rf_fi.feature_importances_
indices = np.argsort(importances)[::-1]
feature_names = X_fi.columns.tolist()
# Plotting the feature importances
plt.figure(figsize=(10,5))
plt.title("Importance of Features")
plt.bar(range(X_fi.shape[1]), importances[indices], align="center",color='brown')
plt.xticks(range(X_fi.shape[1]), [feature_names[i] for i in indices], rotation=45)
plt.xlim([-1, X_fi.shape[1]])
plt.tight_layout()
#save the figure
plt.savefig('./feature_importance.png')
plt.show()


#### Task 4.2: Model Deployment
* Prepare the final model for deployment, ensuring it is optimized and efficient.
* Develop a pipeline for model deployment, including data preprocessing, feature engineering, and prediction.

-------------------------------To be continued in another Jupyter Notebook----------------------------------